### ISSUE TITLE
Fail if no topological sort is possible

### ISSUE THREAD

**Issue by Victorystick** [rollup/rollup#428]
I've been thinking about [a situation](http://rollupjs.org/#%7B%22options%22%3A%7B%22format%22%3A%22es6%22%2C%22moduleName%22%3A%22myBundle%22%2C%22globals%22%3A%7B%7D%2C%22moduleId%22%3A%22%22%7D%2C%22modules%22%3A%5B%7B%22name%22%3A%22main.js%22%2C%22code%22%3A%22export%20%7B%20foo%20%7D%20from%20'.%2Ffoo'%3B%5Cnexport%20%7B%20bar%20%7D%20from%20'.%2Fbar'%3B%22%7D%2C%7B%22name%22%3A%22foo.js%22%2C%22code%22%3A%22import%20%7B%20bar%20%7D%20from%20'.%2Fbar'%3B%5Cn%5Cnexport%20const%20foo%20%3D%20bar%28%20i%20%3D%3E%20i%20%29%3B%22%7D%2C%7B%22name%22%3A%22bar.js%22%2C%22code%22%3A%22import%20%7B%20foo%20%7D%20from%20'.%2Ffoo'%3B%5Cn%5Cnexport%20const%20bar%20%3D%20foo%28%20i%20%3D%3E%20i%20%29%3B%22%7D%5D%7D) (similar to #425) with recursive strong dependencies. There is no topological sort that will make this code run without an immediate `TypeError` failure. We should probably fail and inform the user that we couldn't properly order their code.

```
┌> A -┐
|     |
└- B <┘
```

**Comment by rosskevin** [comment]
Here's my case.

In the babelHelpers_inherits, I'm getting `Uncaught TypeError: Super expression must either be null or a function, not undefined`.  The superclass for `Switch` is not yet defined, even though my `index.js` imports the superclasses in order.  Regardless of my index.js order, rollup sticks Switch's definition (using `iife`) right after the base classes.

[Source is available for perusal](https://github.com/FezVrasta/bootstrap-material-design/tree/v4-dev/js/src)

Here's my index.js

```
import '../../node_modules/babel-polyfill/dist/polyfill'

import Base from './base'
import BaseInput from './baseInput'
import BaseFormControl from './baseFormControl'
import BaseSelection from './baseSelection'
import Util from './util'

import Checkbox from './checkbox'
import CheckboxInline from './checkboxInline'
import CollapseInline from './collapseInline'
import File from './file'
import Radio from './radio'
import RadioInline from './radioInline'
import Select from './select'
import Switch from './switch'
import Text from './text'
import Textarea from './textarea'

import Ripples from './ripples'
import Autofill from './autofill'
import BootstrapMaterialDesign from './bootstrapMaterialDesign'
```
Hierarchy of Switch is:
```
Switch -> Checkbox -> BaseSelection -> BaseInput -> Base
```

This popped up on me after adding what seems to be an unrelated class (it was working prior).  

Since the Switch is the deepest leaf class in my package, it seems strange that it is near the top.

## This case is confirmed to be a topological error.  
Concrete classes all implement a `#rejectMatch`, and pass `invalidComponentMatches: []` to the super.  This `invalidComponentMatches` is iterated and executes `#rejectMatch` on each to try and make sure that users don't use two classes for one input.  The `#rejectMatch` makes sense to be encapsulated in each subclass, but since `import`s are evaluated statically, I have to import the class to put it in the list of components to check/reject.  It would be nice to evaluate these components at runtime, as I know they would be present. All these checks happen after `document.ready`, but I don't see an ES6 way to import/exec a method on a class at runtime.

**Comment by Rich-Harris** [comment]
The sorting behaviour is different as of 0.25, to bring Rollup closer in line with the ES6 module spec. @rosskevin is this still an issue with 0.25?

**Comment by rosskevin** [comment]
The change had no affect on my issue, but again it seems that ES6 forces me into a cycle.  I want each class to carry responsibility for asserting element matches it accepts/rejects, and other classes have to import them to be able to accomplish this, which yields a cycle in class hierarchies.  

TLDR; my code would work in a lazy execution situation, but not the static situation presented via ES6, so I'll have to re-code it from a different perspective, nothing rollup can do to help me.

**Comment by eventualbuddha** [comment]
ES6 modules allow cycles, right @rosskevin? Unless you actually *use* the imported objects as part of module initialization it should be okay.

**Comment by rosskevin** [comment]
@eventualbuddha yes that was my coding error, utilizing it within the constructor code.  I don't see a way for any tool to break that cycle.
